packages/flutter/lib/src/services/platform_protocols.dart

```diff
@@ -0,0 +1,219 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+import 'platform_messages.dart';
+
+import 'package:flutter/foundation.dart';
+
+/// Protocols for platform communication, implemented on top of
+/// [PlatformMessages].
+class PlatformProtocols {
+  /// Returns a [Future] representing an asynchronous invocation of a named
+  /// platform function with JSON-encoded arguments, results, and errors.
+  ///
+  /// If specified, the optional [args] map must contain valid JSON values
+  /// according to [PlatformMessages.sendJSON].
+  ///
+  /// The platform function must be implemented on the native side by a handler
+  /// registered for the specified [channel]. The handler should expect
+  /// invocations using envelopes of the form
+  ///
+  ///     { "name": <function name>,
+  ///       "args": <arguments json>
+  ///     }
+  ///
+  /// If [name] is not specified, [channel] is used in its place.
+  ///
+  /// The native implementation must wrap the result of a successful invocation
+  /// in an envelope of the form
+  ///
+  ///     { "status": "ok",
+  ///       "data": <result>
+  ///     }
+  ///
+  /// Similarly, an error must be wrapped in an envelope of the form
+  ///
+  ///     { "status": "error",
+  ///       "message": <human-readable error message>,
+  ///       "data": <error details>
+  ///     }
+  ///
+  /// In both cases, `data` is optional and treated as `null` when missing.
+  /// The other envelope fields are mandatory. Additional fields, if any, are
+  /// silently discarded.
+  ///
+  /// On successful invocation, the specified [decoder] is used to turn the
+  /// result data into a [T] instance (or `null`), which will then complete
+  /// the returned future. A `null` decoder is treated as a constantly `null`
+  /// function.
+  ///
+  /// On errors, the returned future will complete with a [PlatformException]
+  /// containing the error message and error details data.
+  ///
+  /// Malformed JSON or envelopes, or data that the [decoder] cannot handle,
+  /// will make the returned Future complete with a [FormatException].
+  static Future<T> invokeJSONFunction<T>({
+    @required String channel,
+    String name,
+    Map<String, dynamic> args,
+    T decoder(dynamic json),
+  }) async {
+    final dynamic reply = await PlatformMessages.sendJSON(channel, <String, dynamic>{
+      'name': name ?? channel,
+      'args': args,
+    });
+    return _interpretJSON<T>(reply, decoder);
+  }
```

Comment:
I'll reconsider this method in light of the discussion above about hiding the payload encoding as an implementation detail as well as the discussion below about refactoring `PlatformMessages`. I don't think there should be two `invokeMethod`-like methods in the SDK.

To answer the question: The main difference is the use of the specified JSON envelope for responses, making the success/error status explicit as in HTTP. Beyond that, we use an `args` map rather than an `args` list, basically for the same reason that Dart has named parameters (it also provides a place for the `eventChannel` field in `createJSONBroadcastStream`). And then there's the `decoder` parameter producing a custom `T` rather than a JSON-like value.